packages/hash/shared/src/save.ts

```diff
@@ -1,693 +1,473 @@
+import { isEqual, omit } from "lodash";
 import { ProsemirrorNode, Schema } from "prosemirror-model";
-import { EditorState } from "prosemirror-state";
+import { v4 as uuid } from "uuid";
+import {
+  BlockEntity,
+  isDraftTextContainingEntityProperties,
+  isDraftTextEntity,
+} from "./entity";
 import {
-  getDraftEntityFromEntityId,
   DraftEntity,
   EntityStore,
-  isBlockEntity,
+  getDraftEntityFromEntityId,
   isDraftBlockEntity,
 } from "./entityStore";
 import {
-  BlockEntity,
-  blockEntityIdExists,
-  getTextEntityFromSavedBlock,
-  isDraftTextContainingEntityProperties,
-} from "./entity";
-import {
-  CreateEntityMutation,
-  CreateEntityMutationVariables,
-  CreateEntityTypeSharedMutation,
-  CreateEntityTypeSharedMutationVariables,
+  EntityTypeChoice,
   GetAccountEntityTypesSharedQuery,
   GetAccountEntityTypesSharedQueryVariables,
+  GetPageQuery,
+  GetPageQueryVariables,
   SystemTypeName,
   UpdatePageAction,
   UpdatePageContentsMutation,
   UpdatePageContentsMutationVariables,
+  UpdatePageContentsResultPlaceholder,
 } from "./graphql/apiTypes.gen";
-import {
-  componentNodeToId,
-  EntityNode,
-  findComponentNode,
-  isComponentNode,
-  isEntityNode,
-  textBlockNodeToEntityProperties,
-} from "./prosemirror";
-import { updatePageContents } from "./queries/page.queries";
+import { isEntityNode } from "./prosemirror";
+import { getPageQuery, updatePageContents } from "./queries/page.queries";
+
+type EntityType =
+  GetAccountEntityTypesSharedQuery["getAccountEntityTypes"][number];
 
 /**
- * Our operations need to combine the actions from the previous operation,
- * but the actual code to do this is a bit mundane / messes up code clarity,
- * so I've extracted that to a wrapper function. This means each operation
- * only needs to handle creating their own actions
+ * @todo this assumption of the slug might be brittle,
+ * @todo don't copy from server
  */
-const defineOperation =
-  <T extends any[]>(
-    fn: (
-      entities: BlockEntity[],
-      ...args: T
-    ) => readonly [UpdatePageAction[], BlockEntity[]],
-  ) =>
-  (
-    existingActions: UpdatePageAction[],
-    entities: BlockEntity[],
-    ...args: T
-  ): [UpdatePageAction[], BlockEntity[]] => {
-    const [nextActions, nextEntities] = fn(entities, ...args);
-
-    return [[...existingActions, ...nextActions], nextEntities];
-  };
-
-const removeBlocks = defineOperation(
-  (entities: BlockEntity[], draftBlockEntityIds: DraftEntity["entityId"][]) => {
-    const draftBlockEntityIdsSet = new Set(draftBlockEntityIds);
-
-    const removedBlockEntities = entities
-      .map((block, position) => [block, position] as const)
-      .filter(([block]) => !draftBlockEntityIdsSet.has(block.entityId));
-
-    const updatedEntities = entities.filter(
-      (_, position) =>
-        !removedBlockEntities.some(
-          ([, removedPosition]) => removedPosition === position,
-        ),
-    );
+const capitalizeComponentName = (cId: string) => {
```

Comment:
Agree we should only define this method in one place, but wouldn't `backend-utils` be a more appropriate package for a utility like this?

However I'm not sure about the consequences of adding either `backend-utils` or the `api` as dependencies of `shared` because `frontend` imports `shared`, so perhaps defining this method in `shared` and having the `api` import it from there would make more sense.

Reply:
`save` shouldn't be in shared anymore, so maybe that's part of the solution. It's a legacy thing.